### PR TITLE
Fix for #2947. Added support for case (in)sensitive search.

### LINKED ISSUE #2947: string.StartsWith/EndsWith do not respect StringComparison arg.
This bug hits atleast Firebird and Sqlite providers.

```c#
		[Test, Combinatorial]
		public void SearchStringTest(
			[IncludeDataSources(TestProvName.AllSQLite, TestProvName.AllFirebird)] string context, 
			[Values(nameof(string.StartsWith), nameof(string.EndsWith))] string methodName,
			[Values]bool ignoreCase)
		{
			var stringComparison = ignoreCase ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
			Expression<Func<Model.Person, bool>> GetPredicate(string param) => methodName switch
			{
				nameof(string.StartsWith) => (Model.Person p) => p.LastName.StartsWith(param, stringComparison),
				nameof(string.EndsWith) => (Model.Person p) => p.LastName.EndsWith(param, stringComparison),				
				_ => throw new ArgumentException("Unknown method name " + methodName)
			};

			using (var db = GetDataContext(context))
			{
				Assert.IsTrue(db.Person.Any(GetPredicate("Pupkin")));
				Assert.AreEqual(ignoreCase, db.Person.Any(GetPredicate("pupkin")));
			}			
		}
```

Also there is FB specific issue in FirebirdSqlOptimizer.ConvertSearchStringPredicate. "starting with" is case-sensitive, but linq2db use it in "ignorecase" branch

### Environment details
linq2db version: 3.3
Database Server: Firebird

### PR DESCRIPTION
**Pull request by sdanyliv**
Fixes #2947

### CODE REVIEW

File: Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs [hunk 1]
```diff
@@ -1662,6 +1660,19 @@ ValueTypeExpression ReplaceParameter(IDictionary<Expression,Expression> expressi
 
 		ISqlPredicate ConvertPredicate(IBuildContext? context, Expression expression)
 		{
+			static bool IsIgnoreCase(MethodCallExpression mc)
```
Comment (jack128) [inline Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:1663]: I tried to fix the issue this way, but run into some cache problem. For example 
```
			StringComparison stringComparison;
			Expression<Func<Model.Person, bool>> GetPredicate(string param) => (Model.Person p) => p.LastName.StartsWith(param, stringComparison);
			using (var db = GetDataContext(context))
			{
				stringComparison = StringComparison.CurrentCultureIgnoreCase;
				db.Person.Any(GetPredicate("pupkin"));
				stringComparison = StringComparison.CurrentCulture;
				db.Person.Any(GetPredicate("pupkin")); // 2
			}
```
 linq2db do not call CreateStringPredicate in line 2 (get it from cache???) and search is case-insensitive

Comment (sdanyliv) [reply Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:1663]: Good point, yes probably cache may leverage this. attempt. Will take care in this. part.

**Review by MaceWindu — CHANGES_REQUESTED** [review]

**Comment by MaceWindu** [comment]
For SQL Server 2019 we use case-sensitive database so tests fail. Which rise interesting issue.
I think we shouldn't depend on default match behavior if user specified comparison mode explicitly.

File: Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs (hunk 1 above)
Comment (sdanyliv) [reply Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs]: Resolved.

**Comment by MaceWindu** [comment]
Clarification on what's going on here. This PR updates SQL, generated for Contains/StartsWith/EndsWith methods with `StringComparison` parameter:

- no changes: for calls without parameter we compare as is: using server comparison logic, db or column collation (which one is set)
- when `StringComparison` parameter set we try to generate explicit case-sensitive or case-insensitive comparison (depends on parameter value)

**Review by viceroypenguin — APPROVED** [review]

**Pull request merged**